lib/ruby_lsp/requests/definition.rb

```diff
@@ -79,6 +102,28 @@ def on_command(node)
 
       private
 
+      sig { params(value: String).void }
+      def find_in_index(value)
+        entries = @index.resolve(value, @nesting)
+        return unless entries
+
+        @response = entries.filter_map do |entry|
+          location = entry.location
+          # If the project has Sorbet, then we only want to handle go to definition for constants defined in gems, as an
+          # additional behavior on top of jumping to RBIs. Sorbet can already handle go to definition for all constants
+          # in the project, even if the files are typed false
+          next if HAS_TYPECHECKER && entry.file_path.start_with?(T.must(WORKSPACE_URI.to_standardized_path))
+
+          Interface::Location.new(
+            uri: "file://#{entry.file_path}",
```

Comment:
Yes, I pushed a PR to standardize all our URI handling, but this was an old branch I had around and I missed this while reviewing my own code.

Thanks for pointing out, I switched to using our URI builder.

Reply:
Hmm, I see the standardized URI handling, but I am afraid it doesn't solve all of our problems. I'll create a PR to fix it.

Reply:
Can you elaborate what's missing?

Reply:
Sure. For example:
```ruby
URI.parse("file://" + URI::DEFAULT_PARSER.escape("/foo/bar baz/baz?.rb")).path
# => "/foo/bar%20baz/baz"
```
The escaping we are using isn't comprehensive enough, so it fails to encode things like `?`.

Also, Windows paths use backslashes instead of forward slashes, which means that the URIs we generate for `C:\Windows\System32` would need to do a special treatment to convert it to `file:///C:/Windows/System32`, which I don't think we are doing right now.

Both of these are raising errors in the current `from_path` method:
```ruby
URI::Generic.from_path(path: "/foo/bar baz/quux?.rb")
# /opt/rubies/3.2.0/lib/ruby/3.2.0/uri/generic.rb:770:in `check_path': bad component(expected absolute path component): /foo/bar%20baz/quux?.rb (URI::InvalidComponentError)

URI::Generic.from_path(path: 'c:\windows\system32')
# /opt/rubies/3.2.0/lib/ruby/3.2.0/uri/generic.rb:770:in `check_path': bad component(expected absolute path component): c:%5Cwindows%5Csystem32 (URI::InvalidComponentError)
```

Reply:
For Windows, as far as I know using forward slashes is also valid as paths. And based on this user's report, VS Code does seem to send URIs using forward slashes https://github.com/Shopify/vscode-ruby-lsp/pull/712#issuecomment-1659587130.

FWIW, I think the `URI` module should be handling all of this complexity. I checked the implementation for Rust Analyzer to understand what Windows was supposed to expect, and it turns out everything is handled seamlessly by the URI crate.

Reply:
https://learn.microsoft.com/en-us/archive/blogs/ie/file-uris-in-windows btw

Reply:
I wonder if the core team would be open to change URI, so that we can pass any valid Windows path back and forth and get the right thing. E.g.:

```ruby
uri = URI.build(scheme: "file", path: "C:\whatever\foo.rb") # OK
uri.to_s
# => "file:///C:/whatever/foo.rb"
uri.path
# => "C:\whatever\foo.rb"
```

It would be a valuable contribution as it would make it simpler to support Windows things.

Reply:
It's more complicated than that, since once you get path components, you need to URL-encode them as well, for it to be a valid URI. For example, `C:\Documents and Settings\foo.rb` should map to `file:///C:/Documents%20and%20Settings/foo.rb`.